dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java

```diff
@@ -240,14 +249,114 @@ public Tool refreshTool(final long toolId, final Long userId, final UserDAO user
         final List<Tool> newDBTools = new ArrayList<>();
         newDBTools.add(toolDAO.findById(tool.getId()));
 
+        List<Tag> toolTags;
         // Get tags and update for each tool
-        List<Tag> toolTags = getTags(tool);
+        if (tool.getRegistry().equals(Registry.DOCKER_HUB.toString())) {
+            toolTags = getTagsDockerHub(tool);
+        } else {
+            toolTags = getTags(tool);
+        }
+
         updateTags(toolTags, tool, sourceCodeRepoInterface, tagDAO, fileDAO, toolDAO, fileFormatDAO);
 
         // Return the updated tool
         return newDBTools.get(0);
     }
 
+    public List<Tag> getTagsDockerHub(Tool tool) {
+        final String repo = tool.getNamespace() + '/' + tool.getName();
+        LOG.info(" ======================= Getting tags for: {}================================", tool.getPath());
+
+        final List<Tag> tags = new ArrayList<>();
+
+        Optional<String> dockerHubResponse = getDockerHubToolAsString(tool);
+
+        Map<String, List<Map<String, List<Map<String, String>>>>> manifestDigestMap = new HashMap<>();
```

Comment:
These maps will definitely need commenting to understand what the keys are and what the values are. 
Goes for all three, but especially the nested ones